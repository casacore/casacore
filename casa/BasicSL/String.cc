//# String.cc: String class
//# Copyright (C) 2001,2002,2003
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casa/BasicSL/String.h>

#include <casa/BasicSL/RegexBase.h>
#include <algorithm>
#include <casa/string.h>
#include <casa/sstream.h>

namespace casa { //# NAMESPACE CASA - BEGIN

// Special constructors
String::String(ostringstream &os) {
  *this = os.str();
}

// Count occurrences
Int String::freq(Char c) const {
  size_type p(0);
  Int found(0);
  while (p < length()) {
    if ((p = find(c, p)) == npos) break;
    found++;
    p++;
  }
  return found;
}

Int String::freq(const string &str) const {
  size_type p(0);
  Int found(0);
  while (p < length()) {
    if ((p = find(str, p)) == npos) break;
    found++;
    p++;
  }
  return found;
}

Int String::freq(const Char *s) const {
  size_type p(0);
  Int found(0);
  while (p < length()) {
    if ((p = find(s, p)) == npos) break;
    found++;
    p++;
  }
  return found;
}

Double String::toDouble(const String& string) {
    istringstream instr(string);
    Double var;
    instr >> var;
    if (instr.fail()) {
      var = 0.0;
    }
    return var;
}

Float String::toFloat(const String& string) {
    istringstream instr(string);
    Float var;
    // Initialize in case the string is empty or non-numeric.
    instr >> var;
    if (instr.fail()) {
      var = 0.0;
    }
    return var;
}

Int String::toInt(const String& string) {
    istringstream instr(string);
    Int var;
    // Initialize in case the string is empty or non-numeric.
    instr >> var;
    if (instr.fail()) {
      var = 0.0;
    }
    return var;
}

void String::trim() {
    iterator iter = begin();
    while (iter != end()  &&
           (*iter == ' '  ||  *iter == '\t'  ||
            *iter == '\n' ||  *iter == '\r')) {
        ++iter;
    }
    erase (begin(), iter);
    if (! empty()) {
        iter = end() - 1;
        while (iter != begin()  &&
               (*iter == ' '  ||  *iter == '\t'  ||
                *iter == '\n' ||  *iter == '\r')) {
            --iter;
        }
        ++iter;
        erase (iter, end());
    }
}

// Obtain a (separate) 'sub'-string
SubString String::at(size_type pos, size_type len) {
  return _substr(pos, len);
}

SubString String::at(const string &str, Int startpos) {
  return _substr(index(str, startpos), str.length());
}

SubString String::at(const Char *s, Int startpos) {
  return _substr(index(s, startpos), traits_type::length(s));
}

SubString String::at(Char c, Int startpos) {
  return _substr(index(c, startpos), 1);
}

SubString String::before(size_type pos) {
  return _substr(0, pos);
}

SubString String::before(const string &str, Int startpos) {
  return _substr(0, index(str, startpos));
}

SubString String::before(const Char *s, Int startpos) {
  return _substr(0, index(s, startpos));
}

SubString String::before(Char c, Int startpos) {
  return _substr(0, index(c, startpos));
}

SubString String::through(size_type pos) {
  return _substr(0, pos+1);
}

SubString String::through(const string &str, Int startpos) {
  size_type last(index(str, startpos));
  if (last != npos) last += str.length();
  return _substr(0, last);
}

SubString String::through(const Char *s, Int startpos) {
  size_type last(index(s, startpos));
  if (last != npos) last +=  traits_type::length(s);
  return _substr(0, last);
}

SubString String::through(Char c, Int startpos) {
  size_type last(index(c, startpos));
  if (last != npos) last += 1;
  return _substr(0, last);
}

SubString String::from(size_type pos) {
  return _substr(pos, length()-pos);
}

SubString String::from(const string &str, Int startpos) {
  size_type first(index(str, startpos));
  return _substr(first, length()-first);
}

SubString String::from(const Char *s, Int startpos) {
  size_type first(index(s, startpos));
  return _substr(first, length()-first);
}

SubString String::from(Char c, Int startpos) {
  size_type first(index(c, startpos));
  return _substr(first, length()-first);
}

SubString String::after(size_type pos) {
  return _substr(pos+1, length()-(pos+1));
}

SubString String::after(const string &str, Int startpos) {
  size_type first(index(str, startpos));
  if (first != npos) first += str.length();
  return _substr(first, length()-first);
}

SubString String::after(const Char *s, Int startpos) {
  size_type first(index(s, startpos));
  if (first != npos) first += traits_type::length(s);
  return _substr(first, length()-first);
}

SubString String::after(Char c, Int startpos) {
  size_type first(index(c, startpos));
  if (first != npos) first += 1;
  return _substr(first, length()-first);
}

// Prepend string
void String::prepend(const string &str) {
  insert(size_type(0), str);
}

void String::prepend(const Char *s) {
  insert(size_type(0), s);
}

void String::prepend(Char c) {
  insert(size_type(0), c);
}

// Delete
void String::del(size_type pos, size_type len) {
  erase(pos, len);
}

void String::del(const string &str, size_type startpos) {
  erase(index(str, startpos), str.length());
}

void String::del(const Char *s, size_type startpos) {
  erase(index(s, startpos), traits_type::length(s));
}

void String::del(Char c, size_type startpos) {
  erase(index(c, startpos), 1);
}

// Global substitution
Int String::gsub(const string &pat, const string &repl) {
  Int nmatches(0);
  if (length() == 0 || pat.length() == 0 ||
      length() < pat.length()) return nmatches;
  size_type si(0);
  Int rl(repl.length());
  while (length()-si >= pat.length()) {
    size_type pos = find(pat, si);
    if (pos == npos) break;
    else {
      nmatches++;
      replace(pos, pat.length(), repl);
      si = pos + rl;
    }
  }
  return nmatches;
}

Int String::gsub(const Char *pat, const string &repl) {
  return gsub(String(pat), repl);
}

Int String::gsub(const Char *pat, const Char *repl) {
  return gsub(String(pat), String(repl));
}

// Member utilities
void String::reverse() {
  std::reverse(begin(), end());
}
#if defined(AIPS_SUN_NATIVE)
Int ToUpper(Int a){return toupper(a);}
Int ToLower(Int a){return tolower(a);}
#else
#define ToUpper toupper
#define ToLower tolower
#endif

void String::upcase() {
  std::transform(begin(), end(), begin(), ToUpper);
}

void String::downcase() {
  std::transform(begin(), end(), begin(), ToLower);
}

void String::capitalize() {
  for (iterator p=begin(); p < end(); p++) {
    Bool at_word;
    if ((at_word = islower(*p))) *p = ToUpper(*p);
    else at_word = isupper(*p) || isdigit(*p);
    if (at_word) {
      while (++p < end()) {
        if (isupper(*p)) *p = ToLower(*p);
        else if (!islower(*p) && !isdigit(*p)) break;
      }
    }
  }
}

// RegexBase related functions
String::size_type String::find(const RegexBase &r, size_type pos) const {
  Int unused;
  return r.find(c_str(), length(), unused, pos);
}

String::size_type String::rfind(const RegexBase &r, size_type pos) const {
  Int unused;
  return r.rfind(c_str(), length(), unused, pos-length());
}

Bool String::matches(const string &str, Int pos) const {
  Bool rstat(False);
  if (pos < 0) {
    if (this->index(str,pos) == 0) {
      rstat = True;
      ///    } else {
      ///      cerr << "No Match " << this->index(str, pos) << endl;
    }
  } else {
    if (length() != 0 && str.length() != 0 &&
        length() == pos+str.length() &&
        static_cast<size_type>(pos) < length() &&
        index(str, pos) == static_cast<size_type>(pos)) {
      rstat = True;
    }
  }
  return rstat;
}

Bool String::contains(const RegexBase &r) const {
  Int unused;
  return (r.find(c_str(), length(), unused, 0)) != npos;
}

Bool String::matches(const RegexBase &r, Int pos) const {
  String::size_type l = (pos < 0) ? -pos : length() - pos;
  if (l>length()) return False;
  if (pos<0) return r.match(c_str(), l, 0) == l;
  return r.match(c_str(), length(), pos) == l;
}

String::size_type String::index(const RegexBase &r, Int startpos) const {
  Int unused;
  return r.search(c_str(), length(), unused, startpos);
}

SubString String::at(const RegexBase &r, Int startpos) {
  Int mlen;
  size_type first = r.search(c_str(), length(), mlen, startpos);
  return _substr(first, mlen);
}

SubString String::before(const RegexBase &r, Int startpos) {
  Int mlen;
  size_type first = r.search(c_str(), length(), mlen, startpos);
  return _substr(0, first);
}

SubString String::through(const RegexBase &r, Int startpos) {
  Int mlen;
  size_type first = r.search(c_str(), length(), mlen, startpos);
  if (first != npos) first += mlen;
  return _substr(0, first);
}

SubString String::from(const RegexBase &r, Int startpos) {
  Int mlen;
  size_type first = r.search(c_str(), length(), mlen, startpos);
  return _substr(first, length()-first);
}

SubString String::after(const RegexBase &r, Int startpos) {
  Int mlen;
  size_type first = r.search(c_str(), length(), mlen, startpos);
  if (first != npos) first += mlen;
  return _substr(first, length()-first);
}

void String::del(const RegexBase &r, size_type startpos) {
  Int mlen;
  size_type first = r.find(c_str(), length(), mlen, startpos);
  erase(first, mlen);
}

Int String::gsub(const RegexBase &pat, const string &repl) {
  Int nmatches(0);
  if (length() == 0) return nmatches;
  Int pl;
  size_type si(0);
  Int rl(repl.length());
  while (length() > si) {
    size_type pos = pat.find(c_str(), length(), pl, si);
    if (pos >= npos-1 || pl <= 0) break;
    else {
      nmatches++;
      si = pos + rl;
      if (pos == 0 && si == 0) { 	// could be problem with anchor at begin
	Int pls;
	size_type ps = pat.find(c_str(), length(), pls, pl); // try for begin
	if (ps >= npos-1 || pls <= 0) {
	  replace(pos, pl, repl);	// finish off if no more (anchored) match
	  break;
	}
      }
      // Continue global substitution
      replace(pos, pl, repl);
    }
  }
  return nmatches;
}

// Global functions
String reverse(string str) {
  std::reverse(str.begin(), str.end());
  return str;
}

String upcase(string str) {
  std::transform(str.begin(), str.end(), str.begin(), ToUpper);
  return str;
}

String downcase(string str) {
  std::transform(str.begin(), str.end(), str.begin(), ToLower);
  return str;
}

String capitalize(string str) {
  static_cast<String>(str).capitalize();
  return str;
}

String replicate(Char c, String::size_type n) {
  return String(n, c);
}

String replicate(const string &str, String::size_type n) {
  String t(str);
  t.reserve(n*str.length());
  while (--n > 0) t += str;
  return t;
}

Int split(const string &str, string res[], Int maxn,
	  const string &sep) {
  Int i(0);
  String::size_type pos(0);
  while (i < maxn && pos < str.length()) {
    String::size_type p = str.find(sep, pos);
    if (p == String::npos) p = str.length();
    res[i] = String(str, pos, p-pos);
    i++;
    pos = p + sep.length();
  }
  return i;
}

Int split(const string &str, string res[], Int maxn,
	  const RegexBase &sep) {
  Int i(0);
  String::size_type pos(0);
  Int matchlen;
  while (i < maxn && pos < str.length()) {
    String::size_type p = sep.find(str.c_str(), str.length(), matchlen, pos);
    if (p == String::npos) p = str.length();
    res[i] = String(str, pos, p-pos);
    i++;
    pos = p + matchlen;
  }
  return i;
}

Int split(const string &str, string res[], Int maxn,
	  const Char sep) {
  return split(str, res, maxn, String(sep));
}

String common_prefix(const string &x, const string &y, 
		     Int startpos) {
  if (static_cast<String::size_type>(startpos) == String::npos ||
      static_cast<String::size_type>(startpos) >= x.length() ||
      static_cast<String::size_type>(startpos) >= y.length()) return String();
  String::const_iterator xs(x.begin() + startpos);
  String::const_iterator ys(y.begin() + startpos);
  String::size_type l(0);
  while (xs != x.end() && ys != y.end() && *xs++ == *ys++) l++;
  return String(x, startpos, l);
}

String common_suffix(const string &x, const string &y, 
		     Int startpos) {
  if (startpos >= 0 ||
      startpos + Int(x.length()) < 0 ||
      startpos + Int(y.length()) < 0) return String();
  String::const_iterator xs(x.end() + startpos+1);
  String::const_iterator ys(y.end() + startpos+1);
  String::size_type l(0);
  while (xs != x.begin() && ys != y.begin() && *--xs == *--ys) l++;
  return String(x, x.length()+startpos+1-l, l);
}

String join(string src[], Int n, const string& sep) {
  String x;
  for (Int i=0; i<n; i++) {
    x += src[i];
    if (i != n-1) x += sep;
  }
  return x;
}

Int fcompare(String x, String y) {
  x.downcase();
  y.downcase();
  return x.compare(y);
}

// SubString
SubString &SubString::operator=(const SubString &str) {
  if (this != &str) *this = String(str);
  return *this;
}

SubString &SubString::operator=(const String &str) {
  const_cast<string &>(ref_p).replace(pos_p, len_p, str);
  return *this;
}

SubString &SubString::operator=(const Char *s) {
  const_cast<string &>(ref_p).replace(pos_p, len_p, s);
  return *this;
}

SubString &SubString::operator=(const Char c) {
  const_cast<string &>(ref_p).replace(pos_p, len_p, 1, c);
  return *this;
}

} //# NAMESPACE CASA - END


// Create some needed templates
#if !defined(AIPS_AUTO_STL)

#if defined(__GNUG__)
 template void std::__reverse<char*>(char *, char *, std::random_access_iterator_tag);
#endif

#if defined(AIPS_SUN_NATIVE)
 template void std::__reverse(char *, char *, std::random_access_iterator_tag);
 template std::istream& __rwstd::rw_extract_string ( std::istream&, std::string&, std::char_traits<char>);
#endif


template char *std::transform<char *>(char *, char *, char *, int (*)(int));
#if defined(AIPS_GCC3)
template void std::__reverse<string::iterator>(string::iterator, string::iterator, std::random_access_iterator_tag);
template string::iterator std::transform<string::iterator>(string::iterator, string::iterator, string::iterator, int (*)(int));
#endif

#endif


