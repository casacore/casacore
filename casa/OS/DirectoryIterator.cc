//# DirectoryIterator.cc: Class to define a DirectoryIterator
//# Copyright (C) 1996,2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$


#include <casa/OS/DirectoryIterator.h>
#include <casa/OS/Path.h>
#include <casa/Exceptions/Error.h>

#include <errno.h>                // needed for errno
#include <casa/string.h>               // needed for strerror


namespace casa { //# NAMESPACE CASA - BEGIN

DirectoryIterator::DirectoryIterator()
: itsDirectoryDescriptor (0),
  itsDirectoryEntry      (0),
  itsEnd                 (False),
  itsDirectory           (),
  itsExpression          (".*")
{
    init();
}

DirectoryIterator::DirectoryIterator (const Directory& dir)
: itsDirectoryDescriptor (0),
  itsDirectoryEntry      (0),
  itsEnd                 (False),
  itsDirectory           (dir),
  itsExpression          (".*")
{
    init();
}

DirectoryIterator::DirectoryIterator (const Directory& dir,
				      const Regex& regExpression)
: itsDirectoryDescriptor (0),
  itsDirectoryEntry      (0),
  itsEnd                 (False),
  itsDirectory           (dir),
  itsExpression          (regExpression)
{
    init();
}

DirectoryIterator::DirectoryIterator (const DirectoryIterator& that)
: itsDirectoryDescriptor (0),
  itsDirectoryEntry      (0),
  itsEnd                 (False),
  itsDirectory           (that.itsDirectory),
  itsExpression          (that.itsExpression)
{
    init();
}

DirectoryIterator::~DirectoryIterator()
{
#if defined(AIPS_CRAY_PGI)
    if (itsNameList != 0) {
        for (int i=0; i<itsNrNames; i++) {
	    free(itsNameList[i]);
	}
	free(itsNameList);
    }
#else
    // Free all the memory used by DirectoryIterator.
    closedir (itsDirectoryDescriptor);
#endif
}

DirectoryIterator& DirectoryIterator::operator= (const DirectoryIterator& that)
{
    if (this != &that) {
	closedir (itsDirectoryDescriptor);
	itsDirectoryDescriptor = 0;
	itsDirectoryEntry      = 0;
	itsEnd                 = False;
	itsDirectory           = that.itsDirectory;
	itsExpression          = that.itsExpression;
	init();
    }
    return *this;
}


void DirectoryIterator::init()
{
    // Alas readdir is not supported on the compute nodes of the Cray XT3.
#if defined(AIPS_CRAY_PGI)
    itsNameList = 0;
    itsNrNames = scandir(itsDirectory.path().expandedName().chars(),
			 &itsNameList, 0, alphasort);
    if (itsNrNames < 0) {
        throw AipsError ("DirectoryIterator: error on directory " +
			 itsDirectory.path().expandedName() +
			 ": No such file or directory");
    }
    itsNameInx = 0;
#else
    // Set the private directory on the current working directory
    // Open the directory, if this is not possible throw an exception
    itsDirectoryDescriptor = opendir(itsDirectory.path().expandedName().chars());
    if (itsDirectoryDescriptor == 0){
	throw (AipsError ("DirectoryIterator: error on directory " +
			  itsDirectory.path().expandedName() +
			  ": " + strerror(errno)));
    }
#endif
    // Set itsDirectoryEntry on the first entry.
    operator++();
}

void DirectoryIterator::operator++()
{
    if (itsEnd) {
	throw (AipsError ("DirectoryIterator++ past end on " +
			  itsDirectory.path().expandedName()));
    }
    // Read the entries until a match with the expression is found.
    // Skip . and ..
    String name;
    do {
#if defined(AIPS_CRAY_PGI)
        if (itsNameInx >= itsNrNames) {
	    itsDirectoryEntry = 0;
	} else {
	    itsDirectoryEntry = itsNameList[itsNameInx++];
	}
#else
	itsDirectoryEntry = readdir (itsDirectoryDescriptor);
#endif
	if (itsDirectoryEntry == 0){
	    itsEnd = True;
	    break;
	}
        name = itsDirectoryEntry->d_name;
    }
    while (name == "."  ||  name == ".."
       ||  name.matches (itsExpression) == 0);
}

void DirectoryIterator::operator++(int)
{
    operator++();
}


String DirectoryIterator::name() const
{
    if (itsEnd) {
	throw (AipsError ("DirectoryIterator::name past end on " +
			  itsDirectory.path().expandedName()));
    }
    return itsDirectoryEntry->d_name;
}

File DirectoryIterator::file() const
{
    return itsDirectory.path().expandedName() + "/" + name();
}

void DirectoryIterator::reset()
{
    // Reset the directory to the beginning of the stream
    // and get the first entry.
#if defined(AIPS_CRAY_PGI)
    itsNameInx = 0;
#else
    rewinddir (itsDirectoryDescriptor);
#endif
    itsEnd = False;
    operator++();
}

Bool DirectoryIterator::pastEnd() const
{
    return  (itsEnd == True);
}

} //# NAMESPACE CASA - END

