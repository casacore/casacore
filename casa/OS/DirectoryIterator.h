//# DirectoryIterator.h: Traverse the contents of a directory
//# Copyright (C) 1996
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_DIRECTORYITERATOR_H
#define CASA_DIRECTORYITERATOR_H

//# Includes
#include <casa/aips.h>
#include <casa/OS/File.h>
#include <casa/OS/Directory.h>
#include <casa/Utilities/Regex.h>

#include <dirent.h>                          // needed for DIR


namespace casa { //# NAMESPACE CASA - BEGIN

// <summary>  
// Traverse the contents of a directory
// </summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>

// <use visibility=export>

// <prerequisite> 
//    <li> Basic knowledge of the UNIX file system
//    <li> <linkto class=Directory>Directory</linkto>
//    <li> possibly <linkto class=Regex>Regex</linkto>
// </prerequisite>

// <synopsis>
// DirectoryIterator allows to traverse a directory. In this way all
// file names in a directory can be gotten. Files . and .. will
// always be skipped.
// <p>
// By means of a regular expression it is possible to traverse the
// directory selectively. That is, only the file names matching the regular
// expression will be returned. Note that the regular expression is
// a true regular expression (as defined by class <linkto class=Regex>
// Regex</linkto> and not a file expression as used in shells.
// Thus to get all .cc files, one has to specify ".*\.cc" and not "*.cc".
// <p>
// The <linkto class=File>File</linkto> class can be used to determine if
// a file represents a symlink, directory or regular file.
// </synopsis>

// <example>
// <srcblock>
//    Directory dir("testdir");
//    // Get all .cc files.
//    // Note that Regex is a true regular expression and not a
//    // simplified file expression (like *.cc) as used in shells.
//    DirectoryIterator dirIter(dir, ".*.\cc");
//    while (!dirIter.pastEnd()){
//  	  cout << dirIter.name() << endl;
//	  dirIter++;
//    }
// </srcblock>
// </example>

// <motivation> 
// With this class it is easy to iterate through a directory.
// </motivation>

// <todo asof=$DATE$>
// <li> Allow file expressions like *.cc.
//      However, it's probably better to make that part of Regex.
// </todo>


class DirectoryIterator
{
public:

    // Construct the iterator for the working directory.
    // All entries (except . and ..) will be traversed.
    // It positions the iterator on the first entry.
    DirectoryIterator();

    // Construct the iterator for the given directory.
    // All entries (except . and ..) will be traversed.
    // It positions the iterator on the first entry.
    DirectoryIterator (const Directory& dir);

    // Construct the iterator for the given directory.
    // All entries matching the regular expression will be traversed.
    // It positions the iterator on the first entry.
    DirectoryIterator (const Directory& dir, const Regex& regExpression);

    // Copy constructor (copy semantics).
    // The iterator will be positioned at the beginning.
    DirectoryIterator (const DirectoryIterator& that);

    // Assignment (copy semantics).
    // The iterator will be positioned at the beginning.
    DirectoryIterator& operator= (const DirectoryIterator& that);

    ~DirectoryIterator();

    // Position on the next matching entry in the directory.
    // <br>An exception is thrown if the iterator is already past the end.
    // <group>
    void operator++();
    void operator++(int);
    // </group>

    // Returns the file name at the current position.
    // <br>An exception is thrown if the iterator is already past the end.
    String name() const;

    // Returns a File object for the file at the current position.
    // Note that this adds the path of the directory to get the
    // correct path for the file.
    // <br>An exception is thrown if the iterator is already past the end.
    File file() const;

    // Reposition the directory stream on the first entry.
    void reset();

    // Checks if the iterator is past the end.
    Bool pastEnd() const;

private:
    // Initialize the iterator.
    void init();

    // This variable is used for seeking in the directory.
    // The directory is opened and closed once during the lifetime 
    // of the class.
    DIR *itsDirectoryDescriptor;

    // This structure is used for information of the directory.
    dirent *itsDirectoryEntry;

    // Boolean to check if the directory stream has past the end
    Bool itsEnd;

    // class directory 
    Directory itsDirectory;
    
    // Regular expression if given, with this variable it is possible 
    // to compare files with regular expression.
    Regex itsExpression;

    // Cray XT3 does not support readdir on compute nodes.
    // Use scandir instead.
    dirent** itsNameList;
    int      itsNrNames;
    int      itsNameInx;
};



} //# NAMESPACE CASA - END

#endif
