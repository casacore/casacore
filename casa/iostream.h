//# iostream.h: Interim solution for standard/nonstandard system iostream
//# Copyright (C) 2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_IOSTREAM_H
#define CASA_IOSTREAM_H

//# Define the C standard C++ include file. 
//# This is an interim solution to cater for the SGI non-existence of
//# them (e.g. <cstring>)

//# Make sure any special macros are set
#include <casacore/casa/aips.h>
#include <iostream>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

using std::istream;
using std::ostream;
using std::iostream;
using std::cin;
using std::cout;
using std::cerr;
using std::endl;
using std::flush;
using std::ws;

using std::ios;
using std::streampos;
using std::streamoff;
using std::streamsize;
using std::dec;
using std::hex;
using std::oct;

using std::internal;
using std::left;
using std::right;
using std::fixed;
using std::scientific;

using std::boolalpha;
using std::noboolalpha;
using std::showbase;
using std::noshowbase;
using std::showpoint;
using std::noshowpoint;
using std::showpos;
using std::noshowpos;
using std::skipws;
using std::noskipws;
using std::uppercase;
using std::nouppercase;

} //# NAMESPACE CASACORE - END

#endif
