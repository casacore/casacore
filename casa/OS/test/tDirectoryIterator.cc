//# tDirectoryIterator.cc: Test program for class DirectoryIterator
//# Copyright (C) 1996,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$


#include <casacore/casa/OS/DirectoryIterator.h>
#include <casacore/casa/BasicSL/String.h>
#include <casacore/casa/Utilities/Assert.h>
#include <casacore/casa/Exceptions.h>
#include <casacore/casa/iostream.h>
#include <map>

#include <casacore/casa/namespace.h>
// <summary>
// Test program for class DirectoryIterator.
// </summary>

// This program tests the class DirectoryIterator.
// The results are written to stdout. The script executing this program,
// compares the results with the reference output file.
// <p>
// When an argument is given, no exceptions will be thrown.
// This can be used to check if no memory leaks occur in normal operation.


uInt countFiles (DirectoryIterator& iter)
{
    uInt n=0;
    while (! iter.pastEnd()) {
	n++;
	iter++;
    }
    return n;
}

void doIt (Bool doExcp)
{
    // Define the possible names and if found.
    // This is needed, because in Linux2.4 the order of readdir is not
    // alphabetical (but hash order).
    std::map<String,Bool> usedNames;
    usedNames["a"] = False;
    usedNames["ca"] = False;
    usedNames["ca.cc"] = False;
    String firstName;
    Directory dir("tDirectoryIterator_tmp");
    DirectoryIterator iter(dir, Regex(".*a.*"));
    for (int i=0; i<3; ++i) {
      AlwaysAssertExit (!iter.pastEnd());
      String nm = iter.name();
      if (i == 0) {
	firstName = nm;
      }
      std::map<String,Bool>::iterator uniter = usedNames.find(nm);
      AlwaysAssertExit (uniter != usedNames.end());
      AlwaysAssertExit (uniter->second == False);
      uniter->second = True;
      AlwaysAssertExit (iter.file().path().originalName() ==
		                            "tDirectoryIterator_tmp/" + nm);
      iter++;
    }
    AlwaysAssertExit (iter.pastEnd());
    if (doExcp) {
	try {
	    iter.name();
	} catch (AipsError x) {
	    cout << x.getMesg() << endl;               // past end
	} 
	try {
	    iter++;
	} catch (AipsError x) {
	    cout << x.getMesg() << endl;               // past end
	} 
    }
    iter.reset();
    AlwaysAssertExit (iter.name() == firstName);
    AlwaysAssertExit (countFiles(iter) == 3);

    iter = DirectoryIterator(dir);
    AlwaysAssertExit (countFiles(iter) == 6);
    iter = DirectoryIterator(dir, Regex("."));
    AlwaysAssertExit (countFiles(iter) == 3);
    iter = DirectoryIterator(dir, Regex("c.*"));
    AlwaysAssertExit (countFiles(iter) == 4);
    DirectoryIterator itera(iter);
    AlwaysAssertExit (countFiles(itera) == 4);
    iter = DirectoryIterator(dir, Regex("cc.*"));
    AlwaysAssertExit (countFiles(iter) == 0);
    iter = DirectoryIterator(dir, Regex(".*\\.cc"));
    AlwaysAssertExit (countFiles(iter) == 1);

    if (doExcp) {
	try {
	    DirectoryIterator iter1 (Directory("tDirectoryIterator_tmp/sub"));
	} catch (AipsError x) {
	    cout << x.getMesg() << endl;               // not existing
	} 
    }
}


int main (int argc, const char*[])
{
    try {
	doIt ( (argc<2));
    } catch (AipsError x) {
	cout << "Caught an exception: " << x.getMesg() << endl;
	return 1;
    } 
    cout << "OK" << endl;
    return 0;                           // exit with success status
}
