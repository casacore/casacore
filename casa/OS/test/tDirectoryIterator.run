#! /bin/sh
#-----------------------------------------------------------------------------
# Script to execute the tDirectoryIterator program via assay.
# It defines a few files, etc. to make the program work in any environment.
# Execution:
#     tDirectoryIterator.run     execute using assay (normal procedure)
#     tDirectoryIterator.run 0   only define the subdirectory, etc. (to debug)
#     tDirectoryIterator.run 1   execute as such (to test)
#     tDirectoryIterator.run 1 1 execute without exceptions (to test leakage)
#=============================================================================

/bin/rm -rf tDirectoryIterator_tmp
mkdir tDirectoryIterator_tmp

touch tDirectoryIterator_tmp/a
touch tDirectoryIterator_tmp/b
touch tDirectoryIterator_tmp/c
touch tDirectoryIterator_tmp/c1
touch tDirectoryIterator_tmp/ca
touch tDirectoryIterator_tmp/ca.cc

if [ $# != 0 ]
then
    if [ $1 != 0 ]
    then
        $casa_checktool ./tDirectoryIterator $2
    fi
else
    $casa_checktool ./tDirectoryIterator
fi
