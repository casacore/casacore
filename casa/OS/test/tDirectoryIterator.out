DirectoryIterator::name past end on tDirectoryIterator_tmp
DirectoryIterator++ past end on tDirectoryIterator_tmp
DirectoryIterator: error on directory tDirectoryIterator_tmp/sub: No such file or directory
OK
